src/udp.rs

```diff
@@ -0,0 +1,97 @@
+//! Tokio aware UDP primitives
+
+use io::{Readiness, Ready};
+use reactor::{self, Source};
+use mio::would_block;
+use mio::udp as mio;
+use std::io;
+use std::net::SocketAddr;
+
+/// A UDP socket.
+pub struct UdpSocket {
+    mio: mio::UdpSocket,
+    source: Source,
+}
+
+impl UdpSocket {
+
+    /// Creates a UDP socket from the given address.
+    pub fn bind(addr: &SocketAddr) -> io::Result<UdpSocket> {
+        let mio = try!(mio::UdpSocket::bind(addr));
+        UdpSocket::watch(mio)
+    }
+
+    /// Create and return a new `UdpSocket` backed by the given Mio
+    /// UdpSocket.
+    ///
+    /// This turns an existing socket into a Tokio aware socket.
+    pub fn watch(mio: mio::UdpSocket) -> io::Result<UdpSocket> {
+        let source = try!(reactor::register_source(&mio, Ready::all()));
+
+        Ok(UdpSocket {
+            mio: mio,
+            source: source,
+        })
+    }
+
+    /// Returns the local address of this socket.
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.mio.local_addr()
+    }
+
+    /// Sends data on the socket to the given address. On success, returns the
+    /// number of bytes written.
+    pub fn send_to(&self, buf: &[u8], target: &SocketAddr) -> io::Result<usize> {
```

Comment:
I changed the return type to be more consistent with the TCP implementation. It was `io.Result<Option<usize>`, where a value of `None` means that it WouldBlock.

Does this change makes sense?